packages/aws-cdk-lib/aws-s3/lib/bucket.ts

```diff
@@ -2460,6 +2609,122 @@ export class Bucket extends BucketBase {
     };
   }
 
+  private renderReplicationConfiguration(props: BucketProps): CfnBucket.ReplicationConfigurationProperty | undefined {
+    if (!props.replicationRules || props.replicationRules.length === 0) {
+      return undefined;
+    }
+
+    if (!props.versioned) {
+      throw new Error('Replication rules require versioning to be enabled on the bucket');
+    }
+    if (props.replicationRules.length > 1 && props.replicationRules.some(rule => rule.priority === undefined)) {
+      throw new Error('\'priority\' must be specified for all replication rules when there are multiple rules');
+    }
+    props.replicationRules.forEach(rule => {
+      if (rule.replicationTimeControl && !rule.metrics) {
+        throw new Error('\'replicationTimeControlMetrics\' must be enabled when \'replicationTimeControl\' is enabled.');
+      }
+      if (rule.deleteMarkerReplication && rule.tagFilter) {
+        throw new Error('\'tagFilter\' cannot be specified when \'deleteMarkerReplication\' is enabled.');
+      }
+    });
+
+    const destinationBuckets = props.replicationRules.map(rule => rule.destination.bucket);
+    const kmsKeys = props.replicationRules.map(rule => rule.kmsKey).filter(kmsKey => kmsKey !== undefined) as kms.IKey[];
+
+    const role = new iam.Role(this, 'ReplicationRole', {
+      assumedBy: new iam.ServicePrincipal('s3.amazonaws.com'),
+    });
+
+    // add permissions to the role
+    // @see https://docs.aws.amazon.com/AmazonS3/latest/userguide/setting-repl-config-perm-overview.html
+    role.addToPolicy(new iam.PolicyStatement({
+      actions: ['s3:GetReplicationConfiguration', 's3:ListBucket'],
+      resources: [Lazy.string({ produce: () => this.bucketArn })],
+      effect: iam.Effect.ALLOW,
+    }));
+    role.addToPolicy(new iam.PolicyStatement({
+      actions: ['s3:GetObjectVersionForReplication', 's3:GetObjectVersionAcl', 's3:GetObjectVersionTagging'],
+      resources: [Lazy.string({ produce: () => this.arnForObjects('*') })],
+      effect: iam.Effect.ALLOW,
+    }));
+    if (destinationBuckets.length > 0) {
+      role.addToPolicy(new iam.PolicyStatement({
+        actions: ['s3:ReplicateObject', 's3:ReplicateDelete', 's3:ReplicateTags', 's3:ObjectOwnerOverrideToBucketOwner'],
+        resources: destinationBuckets.map(bucket => bucket.arnForObjects('*')),
+        effect: iam.Effect.ALLOW,
+      }));
+    }
+
```

Comment:
On the KMS topic, I think we could also use the encryption info from the bucket props/context (`encryptionKey`, `bucketKeyEnabled`) and grant decrypt permission. see: https://docs.aws.amazon.com/AmazonS3/latest/userguide/replication-config-for-kms-objects.html#replications

>- You must grant kms:Decrypt permissions for the AWS KMS key that's used to decrypt the source object.
>
>- You must grant kms:Encrypt permissions for the AWS KMS key that's used to encrypt the object replica.

Reply:
I've added grant decrypt:
```ts
    kmsKeys.forEach(kmsKey => {
      kmsKey.grantEncrypt(role);
    });
    this.encryptionKey?.grantDecrypt(role); // added
```
`this.encryptionKey` contains the return value of `parseEncryption`, which takes into account `bucketKeyEnabled` and `encryptionKey`.

And I've also updated integ tests to verify whether encrypted objects can be replicated.

```ts
const sourceKmsKey = new kms.Key(stack, 'SourceKmsKey', {
  removalPolicy: RemovalPolicy.DESTROY,
});

const sourceBucket = new s3.Bucket(stack, 'SourceBucket', {
  removalPolicy: RemovalPolicy.DESTROY,
  versioned: true,
  encryptionKey: sourceKmsKey, // added
  replicationRules: [...]
})
```